test/e2e/helpers.go

```diff
@@ -64,3 +72,126 @@ func verifyHealthStatusMetric(caCtx *test.Context, metricsPath string, metricLab
 	}
 }
+
+func setupServingControlPlaneMetricsRoutes(caCtx *test.Context, serviceMonitorServiceName string) (*v1.Route, error) {
+	routeName := serviceMonitorServiceName
+	metricsRoute := &v1.Route{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      routeName,
+			Namespace: servingNamespace,
+		},
+		Spec: v1.RouteSpec{
+			Port: &v1.RoutePort{
+				TargetPort: intstr.FromString("8444"),
+			},
+			To: v1.RouteTargetReference{
+				Kind: "Service",
+				Name: serviceMonitorServiceName,
+			},
+			TLS: &v1.TLSConfig{
+				Termination: v1.TLSTerminationReencrypt,
+			},
+		}}
+	r, err := caCtx.Clients.Route.Routes(servingNamespace).Create(context.Background(), metricsRoute, meta.CreateOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error creating service monitor route: %v", err)
+	}
+	caCtx.AddToCleanup(func() error {
+		caCtx.T.Logf("Cleaning up service metrics route: %s", routeName)
+		return caCtx.Clients.Route.Routes("openshift-serverless").Delete(context.Background(), routeName, meta.DeleteOptions{})
+	})
+	return r, nil
+}
+
+func verifyServingControlPlaneMetrics(caCtx *test.Context) {
+	serviceMonitorsInstances := []struct {
+		name        string
+		expectedStr string
+	}{{
+		name:        "activator-sm",
+		expectedStr: `activator_client_results{name=""}`,
+	}, {
+		name:        "autoscaler-sm",
+		expectedStr: "autoscaler_actual_pods",
+	}, {
+		name:        "autoscaler-hpa-sm",
+		expectedStr: "hpaautoscaler_client_latency_bucket",
+	}, {
+		name:        "controller-sm",
+		expectedStr: "controller_client_latency_bucket",
+	}, {
+		name:        "domain-mapping-sm",
+		expectedStr: "domainmapping_client_latency_bucket",
+	}, {
+		name:        "domainmapping-webhook-sm",
+		expectedStr: "domainmapping_webhook_client_latency_bucket",
+	}, {
+		name:        "webhook-sm",
+		expectedStr: "webhook_client_latency_bucket",
+	}}
+	for _, sm := range serviceMonitorsInstances {
+		serviceName := sm.name + "-service"
+		_, err := caCtx.Clients.Kube.CoreV1().Services(servingNamespace).Get(context.Background(), serviceName, meta.GetOptions{})
+		if err != nil {
+			caCtx.T.Fatalf("Error getting service monitor service: %v", err)
+		}
+		route, err := setupServingControlPlaneMetricsRoutes(caCtx, serviceName)
+		if err != nil {
+			caCtx.T.Fatalf("Failed to setup operator metrics route: %v", err)
+		}
+		metricsPath := "https://" + route.Spec.Host + "/metrics"
+		metricsURL, err := url.Parse(metricsPath)
+		if err != nil {
+			caCtx.T.Fatalf("Error parsing url for metrics: %v", err)
+		}
+		bToken := getBearerTokenForAuthorizedAccount(caCtx)
+		var reqOption pkgTest.RequestOption = func(request *http.Request) {
+			request.Header.Set("Authorization", fmt.Sprintf("Bearer %s", bToken))
+		}
+		var transportOption spoof.TransportOption = func(transport *http.Transport) *http.Transport {
+			transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+			return transport
+		}
```

Comment:
Could we test this E2E through prometheus as well? After all, it's important for the customer that our metrics end up in prometheus correctly, so they can be queried by Grafana etc and the setup there is critical as prometheus needs to be given the correct RBAC and whatnot. Does that make sense?

Reply:
Rather than "as well" it might be sufficient to **just** test through prometheus if I think about it.

Reply:
There is no Grafana support anymore. I thought about querying Prometheus but that needed some investigation how to access it.

Reply:
we can still pull this in, but I think the test is kind of incomplete since the crucial part (to me as a reviewer) is knowing that the connection between prometheus (the thin scraping the metrics) and the actual metric endpoints is working as intended.

Reply:
One drawback with the Prometheus being used (although it needs to be tested) is that it takes some time to discover targets.